server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java

```diff
@@ -51,23 +51,40 @@ public GradleApiConnector(PreferenceManager preferenceManager) {
   }
 
   /**
-   * Get the Gradle version of the project.
+   * Extracts the GradleVersion for the given project.
+   *
+   * @param projectUri URI of the project to get the gradle version for.
+   * @return Gradle version of the project or empty upon failure.
    */
   public String getGradleVersion(URI projectUri) {
     try (ProjectConnection connection = getGradleConnector(projectUri).connect()) {
-      return getGradleVersion(connection);
+      return getBuildEnvironment(connection).getGradle().getGradleVersion();
     } catch (BuildException e) {
       LOGGER.severe("Failed to get Gradle version: " + e.getMessage());
       return "";
     }
   }
 
-  private String getGradleVersion(ProjectConnection connection) {
-    BuildEnvironment model = connection
+  /**
+   * Extracts the GradleJavaHome for the give project.
+   *
+   * @param projectUri URI of the project to get the gradle java home for.
+   * @return GradleJavaHome of the project or null upon failure.
+   */
+  public File getGradleJavaHome(URI projectUri) {
```

Comment:
This method will return the Java home which used to launch build server itself, do I understand correctly?

Reply:
No, I don't think so. This method gets `GradleJavaHome` for the given gradle project from the [BuildEnvironment](https://docs.gradle.org/current/javadoc/org/gradle/tooling/model/build/BuildEnvironment.html) > [JavaEnvironment](https://docs.gradle.org/current/javadoc/org/gradle/tooling/model/build/JavaEnvironment.html). 

Gradle gets the `GradleJavaHome` property in the following [priority order](https://docs.gradle.org/current/userguide/build_environment.html#priority_for_configurations):
1. Command Line arg
2. Gradle Properties File
3. Environment Variable

In our case only the last two will be valid so, this method will return the `GradleJavaHome` defined in the `gradle.properties` file, otherwise it will fallback to using the `JAVA_HOME` environment variable for the JDK location.

Reply:
Thank you, I learnt from it!

Just a small suggestion: would you like to do some verification about the behavior? Because I found that some time, the behavior from command line is different from calling the TAPI. Just in case...

Reply:
Okay. I will verify it.

Reply:
Continued [here](https://github.com/microsoft/build-server-for-gradle/pull/165#issuecomment-2232644129).